### PR TITLE
[FIX] Template Overriding - Add PR #566 Changes Back

### LINKED ISSUE #719: jinja2.exceptions.TemplateNotFound: toggle-primary-sidebar.html
### Describe the bug

**context**
When I creating the the documentation page of GemGIS (https://gemgis.readthedocs.io/en/latest/), I have recently started to see an error: `jinja2.exceptions.TemplateNotFound: toggle-primary-sidebar.html.`

**expectation**
No error to occur

**bug**
The creating of the docs fails --> https://readthedocs.org/projects/gemgis/builds/20023049/

### Reproduce the bug

The creating of the docs fails --> https://readthedocs.org/projects/gemgis/builds/20023049/

### List your environment

_No response_

### LINKED ISSUE #729: Autosummary and templates not working with versions > 1
### Describe the bug

**context**
Try to build the documentation using autosummary and template overrides to give methods their own pages. Using the same templates as the numpy documentation.

**expectation**
I expected the docs to build properly, each function getting its own page.

**bug**
But instead the docs fail to build.

Here's the first error message:

```console
$ jb build docs
Extension error (pydata_sphinx_theme):
Handler <function update_and_remove_templates at 0x109cf9ee0> for event 'html-page-context' threw an exception (exception: toggle-primary-sidebar.html)
```

**problem**
There is therefore no good way to generate API docs with the latest theme with the nice new searchbar shortcuts and other features. This seems to have come up before:

* #719 (wasn't resolved, just provides a workaround; don't think the issue should have been closed)
* https://github.com/sphinx-doc/sphinx/issues/11315
* https://github.com/executablebooks/jupyter-book/issues/1660
* https://github.com/NREL/floris/issues/635

As mentioned, older versions work fine, but I want *this* version to work.

### Reproduce the bug

1. Clone the sample repository: https://github.com/sequoiap/bookerrors
2. Create a virtual environment: ``python3 -m venv env``
3. Activate the virtual environment: ``source env/bin/activate``
4. Install the project to that virtual environment: ``pip install -e .``
5. Build the docs: ``jb build docs``
6. Bask in the glory of the error message in the terminal.

### List your environment

```
Jupyter Book      : 0.15.1
External ToC      : 0.3.1
MyST-Parser       : 0.18.1
MyST-NB           : 0.17.2
Sphinx Book Theme : 1.0.1
Jupyter-Cache     : 0.6.1
NbClient          : 0.7.4
Sphinx            : 5.3.0
```

### LINKED ISSUE #768: [BUG] PR #691 Broke PR #566
### Describe the bug

As noted in:

- [`pydata-sphinx-theme`: Putting `version-switcher` in Primary Sidebar results in Extension Error #1511 ](https://github.com/pydata/pydata-sphinx-theme/issues/1511)

utilizing template overrides with the `sphinx-book-theme` results in the error:

```
Handler <function update_and_remove_templates at 0x000001A18C5BF700> for event 'html-page-context' threw an exception (exception: toggle-primary-sidebar.html)
```

The root cause was identified as [PR #691](https://github.com/executablebooks/sphinx-book-theme/commit/9e2bc2f63f2637b44c215209a27cb726354a89bd), which reverted the following changes to [`__init__.py`](https://github.com/executablebooks/sphinx-book-theme/blob/master/src/sphinx_book_theme/__init__.py) in [PR #566](https://github.com/executablebooks/sphinx-book-theme/pull/566/files#diff-323e2d703544f60dac082d2b47d8b6a2d8e1594eb3b8f49d569f48f768379252R189):

1. `app.connect("config-inited", update_general_config)` changed to `app.connect("builder-inited", update_general_config)`
2. `update_general_config` parameter `config` removed (`config-inited` events issue two parameters)

The `Sphinx` class (from `sphinx-docs`) calls `config.init_values()` **after** the `sphinx-book-theme` extension is loaded, causing `templates_path` to be overridden and the path to `sphinx_book_theme/theme/sphinx_book_theme/components` lost.

### Reproduce the bug

1. Install `sphinx-book-theme` add add it to `conf.py` `extensions`
2. Run `sphinx-build -b html src build`

Results in the following (see also [`pydata-sphinx-theme` Issue #1511](https://github.com/pydata/pydata-sphinx-theme/issues/1511)):

<details>
<summary><b>traceback:</b></summary>

Traceback (most recent call last):
  File "C:\Users\hendra11\Code\external\poetry_plugin_constrain\.venv\lib\site-packages\sphinx\events.py", line 96, in emit
    results.append(listener.handler(self.app, *args))
  File "C:\Users\hendra11\Code\external\poetry_plugin_constrain\.venv\lib\site-packages\pydata_sphinx_theme\__init__.py", line 224, in update_and_remove_templates
    context[section] = list(filter(_remove_empty_templates, context[section]))
  File "C:\Users\hendra11\Code\external\poetry_plugin_constrain\.venv\lib\site-packages\pydata_sphinx_theme\__init__.py", line 219, in _remove_empty_templates    
    rendered = app.builder.templates.render(tname, context)
  File "C:\Users\hendra11\Code\external\poetry_plugin_constrain\.venv\lib\site-packages\sphinx\jinja2glue.py", line 196, in render
    return self.environment.get_template(template).render(context)
  File "C:\Users\hendra11\Code\external\poetry_plugin_constrain\.venv\lib\site-packages\jinja2\environment.py", line 1010, in get_template
    return self._load_template(name, globals)
  File "C:\Users\hendra11\Code\external\poetry_plugin_constrain\.venv\lib\site-packages\jinja2\environment.py", line 969, in _load_template
    template = self.loader.load(self, name, self.make_globals(globals))
  File "C:\Users\hendra11\Code\external\poetry_plugin_constrain\.venv\lib\site-packages\jinja2\loaders.py", line 126, in load
    source, filename, uptodate = self.get_source(environment, name)
  File "C:\Users\hendra11\Code\external\poetry_plugin_constrain\.venv\lib\site-packages\sphinx\jinja2glue.py", line 217, in get_source
    raise TemplateNotFound(template)
jinja2.exceptions.TemplateNotFound: toggle-primary-sidebar.html

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\hendra11\Code\external\poetry_plugin_constrain\.venv\lib\site-packages\sphinx\cmd\build.py", line 285, in build_main
    app.build(args.force_all, args.filenames)
  File "C:\Users\hendra11\Code\external\poetry_plugin_constrain\.venv\lib\site-packages\sphinx\application.py", line 353, in build
    self.builder.build_update()
  File "C:\Users\hendra11\Code\external\poetry_plugin_constrain\.venv\lib\site-packages\sphinx\builders\__init__.py", line 311, in build_update
    self.build(to_build,
  File "C:\Users\hendra11\Code\external\poetry_plugin_constrain\.venv\lib\site-packages\sphinx\builders\__init__.py", line 378, in build
    self.write(docnames, list(updated_docnames), method)
  File "C:\Users\hendra11\Code\external\poetry_plugin_constrain\.venv\lib\site-packages\sphinx\builders\__init__.py", line 586, in write
    self._write_serial(sorted(docnames))
  File "C:\Users\hendra11\Code\external\poetry_plugin_constrain\.venv\lib\site-packages\sphinx\builders\__init__.py", line 596, in _write_serial
    self.write_doc(docname, doctree)
  File "C:\Users\hendra11\Code\external\poetry_plugin_constrain\.venv\lib\site-packages\sphinx\builders\html\__init__.py", line 672, in write_doc
    self.handle_page(docname, ctx, event_arg=doctree)
  File "C:\Users\hendra11\Code\external\poetry_plugin_constrain\.venv\lib\site-packages\sphinx\builders\html\__init__.py", line 1076, in handle_page
    newtmpl = self.app.emit_firstresult('html-page-context', pagename,
  File "C:\Users\hendra11\Code\external\poetry_plugin_constrain\.venv\lib\site-packages\sphinx\application.py", line 481, in emit_firstresult
    return self.events.emit_firstresult(event, *args,
  File "C:\Users\hendra11\Code\external\poetry_plugin_constrain\.venv\lib\site-packages\sphinx\events.py", line 117, in emit_firstresult
    for result in self.emit(name, *args, allowed_exceptions=allowed_exceptions):
  File "C:\Users\hendra11\Code\external\poetry_plugin_constrain\.venv\lib\site-packages\sphinx\events.py", line 107, in emit
    raise ExtensionError(__("Handler %r for event %r threw an exception") %
sphinx.errors.ExtensionError: Handler <function update_and_remove_templates at 0x000001B8B89F7AF0> for event 'html-page-context' threw an exception (exception: toggle-primary-sidebar.html)

Extension error (pydata_sphinx_theme):
Handler <function update_and_remove_templates at 0x000001B8B89F7AF0> for event 'html-page-context' threw an exception (exception: toggle-primary-sidebar.html)
</details>

### List your environment

OS: `Windows 10, 20H2`
python: `3.8.10`
Sphinx: `6.2.1`
jinja2: `3.1.2`
sphinx-book-theme: `1.0.1`
pydata-sphinx-theme: `0.14.1`

### PR DESCRIPTION
**Pull request by adam-grant-hendry**
- Change `update_general_config()` to run on `config-inited` event in `setup()` method
- Add `config` parameter back to `update_general_config()` method

Fixes: 
- fixes #768 
- fixes #729 
- fixes #719

### CODE REVIEW

**Comment by adam-grant-hendry** [comment]
@choldgraf Do you have an option for me to request a reviewer?

File: src/sphinx_book_theme/__init__.py [hunk 1]
```diff
@@ -235,7 +235,7 @@ def setup(app: Sphinx):
     app.connect("builder-inited", check_deprecation_keys)
     app.connect("builder-inited", update_sourcename)
     app.connect("builder-inited", update_context_with_repository_info)
-    app.connect("builder-inited", update_general_config)
+    app.connect("config-inited", update_general_config)
```
Comment (choldgraf) [inline src/sphinx_book_theme/__init__.py:238]: I seem to remember that the reason we did this is because the `config-inited` event actually occurs _before_ the theme is initialized (or something like this) so it wasn't actually being triggered in the way that we thought. Does this make sense and / or are you reasonably certain that we won't run into this issue here?

Comment (adam-grant-hendry) [reply src/sphinx_book_theme/__init__.py:238]: @choldgraf I am certain. Please see my work in [pydata-sphinx-theme #1511](https://github.com/pydata/pydata-sphinx-theme/issues/1511#issuecomment-1767254825). It happens after.

Comment (choldgraf) [reply src/sphinx_book_theme/__init__.py:238]: Just to confirm - in that issue you mention putting `sphinx_book_theme` in `extensions` as in `extensions = [..., "sphinx_book_theme", ]`. I don't think that's something we want to _require_ for users, we just want it to be listed in `html_theme`[^1]. Does this fix require the book theme to be listed in extensions?

Apologies for the nit-picking here, I've just been burned by the confusing nature of Sphinx events and am inherently distrustful of them 😅

[^1]: For example here it is for this theme's docs:

    https://github.com/executablebooks/sphinx-book-theme/blob/1b6e9eb7496bfedff084a304cd2abe2cb46bd7f5/docs/conf.py#L18-L36

Comment (adam-grant-hendry) [reply src/sphinx_book_theme/__init__.py:238]: > Just to confirm - in that issue you mention putting sphinx_book_theme in extensions as in extensions = [..., "sphinx_book_theme", ]. I don't think that's something we want to require for users, we just want it to be listed in html_theme[1](https://github.com/executablebooks/sphinx-book-theme/pull/770#user-content-fn-1-edb9af0702d9ae3dcab46282b225bd4c). Does this fix require the book theme to be listed in extensions?

Oh wow, that was not clear at all, but see now it was alluded to in [#1660](https://github.com/executablebooks/jupyter-book/issues/1660#issue-1162755546):

> I have narrowed the problem down to the fact that sphinx_book_theme, in addition to being used as the html_theme, is also included as an extension. I'm not too familiar with Sphinx, and am unsure why and if a theme should be included as an extension. However, the [documentation for sphinx_book_theme](https://sphinx-book-theme.readthedocs.io/en/latest/tutorials/get-started.html#) seems to indicate that it is sufficient with html_theme="sphinx_book_theme".

Yes, removing it from `extensions` fixes my problem, but since this has bitten several people downstream:

- [Autosummary and templates not working with versions > 1](https://github.com/executablebooks/sphinx-book-theme/issues/729)
- [jinja2.exceptions.TemplateNotFound: toggle-primary-sidebar.html](https://github.com/sphinx-doc/sphinx/issues/11315)
- [https://github.com/executablebooks/jupyter-book/issues/1660](https://github.com/NREL/floris/issues/635)

1. Could we add a warning note to the docs that this package is a theme only, not an extension?

The error that is thrown points to `pydata-sphinx-theme`, which is misleading and unhelpful.

2. If this package is also to be used as an extension, should `__init__.py::setup()` be modified to trigger on **both** `builder-inited` **and** `config-inited`?

Comment (adam-grant-hendry) [reply src/sphinx_book_theme/__init__.py:238]: > I seem to remember that the reason we did this is because the config-inited event actually occurs before the theme is initialized (or something like this) so it wasn't actually being triggered in the way that we thought.

@choldgraf Switching back to `config-inited` isn't breaking any tests...do you have a regression test in place somewhere?

Comment (jayaddison) [reply src/sphinx_book_theme/__init__.py:238]: @adam-grant-hendry nice work tracking this down - I've arrived here from your [explanatory comment](https://github.com/executablebooks/sphinx-book-theme/issues/719#issuecomment-1791000570) on the `gemgis` thread.

My attempt to trace the current status and archaeology of this:

* Current status: good news: not many documentation sites have `sphinx-book-theme` in their extensions; so although this has affected a few folks, it appears to me (based on [this approximate GitHub code search](https://github.com/search?q=path%3A%2Fconf.py%24%2F+extensions+%2F%5Cbextensions.*%3D%2F+sphinx_book_theme+%2Fsphinx_book_theme.*%2C%2F+%24&type=code)) that it's relatively limited.  I'll do a bit more searching around and may open a few pull requests as long as I don't find many codebases affected.

* Archaeology: confusing news: my guess is that this was introduced to `gemgis` during a theme migration from `sphinx-rtd-theme`, which intentionally _does_ state -- in contrast -- that it [_should_ be added as an extension](https://sphinx-rtd-theme.readthedocs.io/en/stable/installing.html#how-to-install-and-use-the-theme).  Does it seem unusual that `sphinx-rtd-theme` has to do that?

Comment (jayaddison) [reply src/sphinx_book_theme/__init__.py:238]: @adam-grant-hendry also a potential puzzler / counterexample: I'm not currently able to replicate this failure when building the `deephyper` documentation (using Sphinx 6.2.1 [from this spec](https://github.com/deephyper/deephyper/blob/92b8bf9d0538b5b39a7e961f2cefedb64fdc9300/setup.py#L86) and with [`sphinx_book_theme` v1.0.1](https://github.com/deephyper/deephyper/blob/92b8bf9d0538b5b39a7e961f2cefedb64fdc9300/setup.py#L87) included in the [Sphinx `extensions` configuration](https://github.com/deephyper/deephyper/blob/92b8bf9d0538b5b39a7e961f2cefedb64fdc9300/docs/conf.py#L79)).

Comment (choldgraf) [reply src/sphinx_book_theme/__init__.py:238]: Is there a way for us to resolve this in a way that:

- Does not require users to put `sphinx_book_theme` in extensions (because many people don't think to do this in general, and it seems redundant from setting it as a theme)
- Will work the same whether `sphinx_book_theme` is in extensions or not (because some folks will nonetheless use the theme in their extensions list)

Comment (adam-grant-hendry) [reply src/sphinx_book_theme/__init__.py:238]: > Is there a way for us to resolve this

Yes, just add a warning note to your documentation that you should not add it to to `extensions`. I think it is as simple as that.

Comment (adam-grant-hendry) [reply src/sphinx_book_theme/__init__.py:238]: > Does it seem unusual that `sphinx-rtd-theme` has to do that?

No because of the note they add afterwards that explains "Adding this theme as an extension is what enables localization of theme strings in your translated output."

An extension should modify an aspect of the document processing whereas a theme should modify the appearance of the (typically HTML) output.

Comment (choldgraf) [reply src/sphinx_book_theme/__init__.py:238]: OK makes sense - so does that mean that:

> just add a warning note to your documentation that you should not add it to extensions. I think it is as simple as that

Would then supercede this PR?

**Comment by agoose77** [comment]
@adam-grant-hendry thanks for opening this PR — I'm picking up the thread to try and get the issue resolved. Could you confirm whether your last reply to @choldgraf above is a reasonable solution - to simply inform users that they should not place the theme in `extensions`? If so, then we can close *this* PR? :)

**Comment by ghisvail** [comment]
I am also affected by the `html-page-context` event [error](https://github.com/medkit-lib/medkit/pull/7). It was a bot hard to follow this thread, just wanted to know how close you think you are to a working fix, and possibly a follow-up release ?

**Comment by choldgraf** [comment]
@ghisvail the short answer is to just not add this theme to your `extensions` list. Adding it as the `html_theme` is enough and should avoid this bug.

**Comment by adam-grant-hendry** [comment]
@agoose77 Yes, that should be fine. A new PR can be created to add the note to the documentation.

**Comment by ghisvail** [comment]
> @ghisvail the short answer is to just not add this theme to your `extensions` list. Adding it as the `html_theme` is enough and should avoid this bug.
> 
> [#770 (comment)](https://github.com/executablebooks/sphinx-book-theme/pull/770#discussion_r1394965818)

I am sorry to insist, but I don't understand what you mean. I am using JB, and the only reference to `extensions` is here:

```yaml
  extra_extensions:
  - 'sphinx.ext.autodoc'
  - 'sphinx.ext.autosummary'
  - 'sphinx.ext.napoleon'
  - 'sphinx.ext.viewcode'
  - 'sphinx_toolbox.more_autosummary'
  - 'sphinx_toolbox.more_autodoc.typehints'
  - 'sphinxcontrib.mermaid'
```

I don't see any explicit reference to `sphinx-book-theme`.

**Comment by ghisvail** [comment]
Now I got it, it's `jupyter-book config` which adds `sphinx-book-theme` to `extensions` in the generated `docs/conf.py`.

If I remove it manually and then perform a build with `sphinx-build`, then it works.
